### ISSUE TITLE
fim tokens not always embedded in prompt + top_p

### ISSUE THREAD

**Issue by ryanolson** [huggingface/llm-vscode#29]
## Observation

I noticed that the vs code request uses `top_p` set to `0.95`, but the default curl command to the latest inference server defaults that value to `None`.  This seems to have a negative effect on the performance of the starcoder model.

Similarly, I noticed that the fim start/middle/suffix tokens are not always included in the prompt.

## Possible Solutions

- Always use fim tokens
- Change the default value of `top_p` to `null` or at least expose it as a parameter
- Enable more options in the extensions, specifically,
  - stop tokens -- it's nice to add `\n` as a stop token so you only generate a single line
  - max_new_tokens -- set to line width / 120-ish

Single line completions feel a bit better, which could be accomplished using a single `\n` stop token.  If we could have use multi-tab or a new hot key, then multi-line would feel better.  That is, use tab to complete the current line; multiple tabs will subsequently complete another line of the suggested auto-complete, and finally a special hot key to accept the full multi-line suggestion.

## Details

In VS Code if I have a simple python file

```

    def test():
        x=1+1
        assert x == 
```

I'm seeing the following request on the server; note: the fim tokens are not included:
```
2023-05-10T22:34:38.711415Z  INFO HTTP request{otel.name=POST /generate http.client_ip= http.flavor=1.1 http.host=a100:8080 http.method=POST http.route=/generate http.scheme=HTTP http.target=/generate http.user_agent=node-fetch otel.kind=server trace_id=44d3a824707eb2d84d040afb9fa8ac59}:generate{req=Json(GenerateRequest { inputs: "\n    def test():\n        x=1+1\n        assert x == ", parameters: GenerateParameters { best_of: None, temperature: None, repetition_penalty: None, top_k: None, top_p: Some(0.95), typical_p: None, do_sample: false, max_new_tokens: 60, return_full_text: None, stop: ["<|endoftext|>"], truncate: None, watermark: false, details: false, seed: None } }) total_time="33.124231ms" validation_time="438.753µs" queue_time="80.04µs" inference_time="32.605648ms" time_per_token="32.605648ms" seed="Some(9473460394149812587)"}: text_generation_router::server: router/src/server.rs:285: Output: 
```

The corresponding output on the console is:
```
INPUT to API: (with parameters {"max_new_tokens":60,"temperature":null,"do_sample":false,"top_p":0.95,"stop":["<|endoftext|>"]}) 

    def test():
        x=1+1
        assert x == 
OUTPUT from API:

```

The output is a `\n`.

Now if I make the call to the server using `curl` with the fim tokens in place (note, i'm adding an extra stop token):
```
deepops@a100:~$ time curl 127.0.0.1:8080/generate     -X POST     -d '{"inputs":"<fim_prefix>\n    def test():\n        x=1+1\n        assert x == <fim_middle><fim_suffix>","parameters":{"max_new_tokens":60,"stop":["<|endoftext|>","\n"]}}'     
-H 'Content-Type: application/json'
{"generated_text":"  def test():\n        x=1+1\n        assert x == 2\n"}
```

I see the corresponding server-side output:
```
2023-05-10T22:12:14.591559Z  INFO HTTP request{otel.name=POST /generate http.client_ip= http.flavor=1.1 http.host=127.0.0.1:8080 http.method=POST http.route=/generate http.scheme=HTTP http.target=/generate http.user_agent=curl/7.68.0 otel.kind=server trace_id=9b28be13203a06e56e54cfea2cf21d21}:generate{req=Json(GenerateRequest { inputs: "<fim_prefix>\n    def test():\n        x=1+1\n        assert x == <fim_middle><fim_suffix>", parameters: GenerateParameters { best_of: None, temperature: None, repetition_penalty: None, top_k: None, top_p: None, typical_p: None, do_sample: false, max_new_tokens: 60, return_full_text: None, stop: ["<|endoftext|>", "\n"], truncate: None, watermark: false, details: false, seed: None } }) total_time="448.996747ms" validation_time="353.56µs" queue_time="66.814µs" inference_time="448.576563ms" time_per_token="26.386856ms" seed="None"}: text_generation_router::server: router/src/server.rs:285: Output:   def test():
        x=1+1
        assert x == 2

```

Note: `top_p: 0.95` when issued via curl, which is the default setting of the latest version of the `text-generation-inference` server

```
deepops@a100:~$ time curl 127.0.0.1:8080/generate     -X POST     -d '{"inputs":"\n    def test():\n        x=1+1\n        assert x == ","parameters":{"max_new_tokens":60,"stop":["<|endoftext|>"],"top_p":0.95}}'     -H 'Content-Type: applicatio
n/json'
{"generated_text":""}
```

server-side:
```
2023-05-10T22:37:01.085903Z  INFO HTTP request{otel.name=POST /generate http.client_ip= http.flavor=1.1 http.host=127.0.0.1:8080 http.method=POST http.route=/generate http.scheme=HTTP http.target=/generate http.user_agent=curl/7.68.0 otel.kind=server trace_id=b72534db620649bd7c296957040cb18c}:generate{req=Json(GenerateRequest { inputs: "\n    def test():\n        x=1+1\n        assert x == ", parameters: GenerateParameters { best_of: None, temperature: None, repetition_penalty: None, top_k: None, top_p: Some(0.95), typical_p: None, do_sample: false, max_new_tokens: 60, return_full_text: None, stop: ["<|endoftext|>"], truncate: None, watermark: false, details: false, seed: None } }) total_time="32.852386ms" validation_time="332.854µs" queue_time="60.754µs" inference_time="32.458948ms" time_per_token="32.458948ms" seed="Some(14293755121722667098)"}: text_generation_router::server: router/src/server.rs:285: Output: 
```

**Comment by ryanolson** [comment]
One more comment regarding auto complete. 

In the previous comment I suggested that single line auto completes feel right. I will extend that to say that the auto complete suggestion should be for the natural scope of the suggestion. 

So if the suggestion was for a for loop, it makes sense to suggest the entire for loop. 

What I found odd about the behavior of the combination of extension + backend model is that it would often (in my limited tests) complete the current line correctly but then go on to the next line and start a new class. I found that I often had to delete the tail of the auto complete. 

Not sure if you do that in the trained model or if you use language semantics and place your fim tokens to denote scope boundaries.

**Comment by McPatate** [comment]
`top_p` and`fim` issues are fixed in the new upcoming `0.0.39`.

As for the repetition you're observing, I think postprocessing is the way to go. There is a lot of work to do on making suggestions more useful. We're working on it, please keep the feedback coming as the extensions evolve!
